docs/articles/reference/upgrading-to-v16.md

```diff
@@ -0,0 +1,264 @@
+## Upgrading to v0.16
+
+Bytewax v0.16 introduces major changes to the way that custom Bytewax inputs are created, as well as improvements to and windowing and execution. In this article, we'll cover some of the updates we've made to our API to make upgrading to v0.16 easier.
+
+## Input changes
+
+In v0.16, we have restructured input to be based around partitions in order to support rescaling stateful dataflows in the future. We have also dropped the `Config` suffix to our input classes. As an example, `KafkaInputConfig` has been renamed to `KafkaInput` and has been moved to `bytewax.connectors.kafka.KafkaInput`.
+
+`ManualInputConfig` is now replaced by two base superclasses that can be subclassed to create custom input sources. They are `DynamicInput` and `PartitionedInput`.
+
+### `DynamicInput`
+
+`DynamicInput` sources are input sources that support reading distinct items from any number of workers concurrently.
+
+`DynamicInput` sources do not support resume state and thus generally do not provide delivery guarantees better than at-most-once.
+
+### `PartitionedInput`
+
+`PartitionedInput` sources can keep internal state on the current position of each partition. If a partition can "rewind" and resume reading from a previous position, they can provide delivery guarantees of at-least-once or better.
```

Comment:
by `or better`, do we mean exactly once with some caveats? Maybe we could be more specific here